input/system/rds/logs.go

```diff
@@ -177,3 +122,119 @@ ErrorCleanup:
+	}
+
+	return identifier, nil
+}
+
+func downloadRdsLogFilePortion(rdsSvc *rds.RDS, tmpFile *os.File, logger *util.Logger, identifier *string, logFileName *string, lastMarker *string) (newBytesWritten int, newMarker *string, additionalDataPending bool, err error) {
+	var resp *rds.DownloadDBLogFilePortionOutput
+	resp, err = rdsSvc.DownloadDBLogFilePortion(&rds.DownloadDBLogFilePortionInput{
+		DBInstanceIdentifier: identifier,
+		LogFileName:          logFileName,
+		Marker:               lastMarker, // This is not set for the initial call, so we only get the most recent lines
+	})
+
+	if err != nil {
+		err = fmt.Errorf("Error downloading logs: %s", err)
+		return
+	}
+
+	if resp.LogFileData == nil {
+		logger.PrintVerbose("Rds/Logs: No log data in response, skipping")
+		return
+	}
+
+	if len(*resp.LogFileData) > 0 {
+		_, err = tmpFile.WriteString(*resp.LogFileData)
+		if err != nil {
+			err = fmt.Errorf("Error writing to tempfile: %s", err)
+			return
+		}
+		newBytesWritten += len(*resp.LogFileData)
+	}
+
+	newMarker = resp.Marker
+	additionalDataPending = *resp.AdditionalDataPending
+
+	return
+}
+
+// Analyze and submit at most the trailing 10 megabytes of the retrieved RDS log file portions
+//
+// This avoids an OOM in two edge cases:
+// 1) When starting the collector, as we always load the last 10,000 lines (which may be very long)
+// 2) When extremely large values are output in a single log event (e.g. query parameters in a DETAIL line)
+//
+// We intentionally throw away data here (and warn the user about it), since the alternative
+// is often a collector crash (due to OOM), which would be less desirable.
+const maxLogParsingSize = 10 * 1024 * 1024
+
+func readLogFilePortion(tmpFile *os.File, bytesWritten int, logger *util.Logger) ([]byte, *os.File, error) {
+	exceededMaxParsingSize := bytesWritten > maxLogParsingSize
+	if exceededMaxParsingSize {
+		logger.PrintWarning("RDS log file portion exceeded more than 10 MB of data in 30 second interval, collecting most recent data only (skipping %d bytes)", bytesWritten-maxLogParsingSize)
+
+		_, err := tmpFile.Seek(int64(bytesWritten-maxLogParsingSize), io.SeekStart)
+		if err != nil {
+			return nil, tmpFile, fmt.Errorf("Error seeking tempfile: %s", err)
+		}
+	}
+
+	// Read the data into memory for analysis
+	buf := make([]byte, bytesWritten)
+	_, err := io.ReadFull(tmpFile, buf)
+	if err != nil {
+		return nil, tmpFile, fmt.Errorf("Error reading %d bytes from tempfile: %s", len(buf), err)
+	}
+
+	// If necessary, recreate tempfile with just the data we're analyzing
+	// (this supports the later read of the temp file during the log upload)
+	if exceededMaxParsingSize {
+		truncatedTmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			return nil, tmpFile, fmt.Errorf("Error allocating tempfile for logs: %s", err)
+		}
+
+		_, err = truncatedTmpFile.Write(buf)
+		if err != nil {
+			truncatedTmpFile.Close()
+			os.Remove(truncatedTmpFile.Name())
```

Comment:
There's not much to do with the `Close` and `Remove` errors at this points, so I assume you're intentionally not even logging them?

Reply:
Yeah, fair question. I wonder if there is a case where these error out and it's actually useful to report them (I can't really imagine why this should fail).

That said, we've seen leftover temp files - and they might be something that we can understand better if this has extra error checks. Will work on adding them.

Reply:
@msakrejda I've added logging for the temporary file close/removal - note I've applied this across the board for consistency, not just in the parts that were changed here.